src/ai/backend/manager/api/session.py

```diff
@@ -2178,6 +2180,111 @@ async def shutdown_service(request: web.Request, params: Any) -> web.Response:
     return web.Response(status=204)
 
 
+async def find_dependency_sessions(
+    root_session_name_or_id: Union[uuid.UUID, str],
+    db_connection: SAConnection,
+    access_key: AccessKey,
+):
+    dependency_session_cache: Dict[str, List] = {}
+
+    async def _find_dependency_sessions(session_name_or_id: Union[uuid.UUID, str]):
+        session_infos = await match_session_ids(
+            session_name_or_id,
+            access_key=access_key,
+            db_connection=db_connection,
+        )
+
+        assert len(session_infos) >= 1, "session not found!"
+
+        session_id = str(session_infos[0].get("session_id"))
+        session_name = session_infos[0].get("session_name")
+
+        assert isinstance(session_id, get_args(Union[uuid.UUID, str]))
+        assert isinstance(session_name, str)
+
+        session_dependency_query = (
+            sa.select(
+                [
+                    session_dependencies.c.depends_on,
+                ]
+            )
+            .select_from(session_dependencies)
+            .where(session_dependencies.c.session_id == session_id)
+        )
+
+        kernel_query = (
+            sa.select(
+                [
+                    kernels.c.status,
+                    kernels.c.status_changed,
+                ]
+            )
+            .select_from(kernels)
+            .where(kernels.c.session_id == session_id)
+        )
+
+        session_dependency_query_result, kernel_query_result = list(
+            map(
+                lambda result: result.mappings().all(),
+                await asyncio.gather(
+                    db_connection.execute(session_dependency_query),
+                    db_connection.execute(kernel_query),
+                ),
+            )
+        )
+
+        dependency_session_ids = list(
+            map(lambda x: str(x.get("depends_on")), session_dependency_query_result)
+        )
+
+        session_info: Dict[str, Union[List, str]] = {
+            "session_id": session_id,
+            "session_name": session_name,
+            "status": str(kernel_query_result[0].get("status")),
+            "status_changed": str(kernel_query_result[0].get("status_changed")),
+            "depends_on": [],
+        }
+
+        if dependency_session_cache.get(session_id):
+            session_info["depends_on"] = dependency_session_cache[session_id]
+            return session_info
+
+        session_info["depends_on"] = await asyncio.gather(
+            *map(
+                lambda dependency_session_id: asyncio.create_task(
+                    _find_dependency_sessions(dependency_session_id)
+                ),
+                dependency_session_ids,
+            )
+        )
```

Comment:
Is there an advantage to creating and gathering multiple tasks? It'll make exception handling trickier. Also, if you have specific reason to do so, please add ways to handle potential exceptions which can be raised on `_find_dependency_sessions()`.

Reply:
I thought `db_connection.execute` is a relatively high-cost operation and should be executed recursively in this function, so I wanted them not to block the Backend.ai manager's other operations while running a bunch of `db_connection.execute`.

But actually, I'm a little confused about python asyncio model yet. For that purpose, is forking a thread and running the operations from that thread separately better? Is this code cannot ensure concurrency?